node/src/exchange/server.rs

```diff
@@ -1,52 +1,408 @@
+
+        let received = select! {
+            _ = poll_fn(move |cx| handler.poll(cx, &mut sender)) => panic!("shouldn't return"),
+            r = receiver => { r.unwrap() }
+        };
+
+        assert_eq!(received.len(), 1);
+        assert_eq!(received[0].status_code, i32::from(StatusCode::Invalid));
+    }
+
+    #[tokio::test]
+    async fn request_hash_test() {
+        let (store, _) = gen_filled_store(1);
+        let stored_header = store.get_head().unwrap();
+        let mut handler = ExchangeServerHandler::new(Arc::new(store));
+        let (mut sender, receiver) = create_test_response_sender();
+
+        handler.on_request_received(
+            PeerId::random(),
+            "test",
+            HeaderRequest::with_hash(stored_header.hash()),
+            (),
+        );
+
+        let received = select! {
+            _ = poll_fn(move |cx| handler.poll(cx, &mut sender)) => panic!("shouldn't return"),
+            r = receiver => { r.unwrap() }
+        };
+
+        assert_eq!(received.len(), 1);
+        assert_eq!(received[0].status_code, i32::from(StatusCode::Ok));
+        let decoded_header = ExtendedHeader::decode(&received[0].body[..]).unwrap();
+        assert_eq!(decoded_header, stored_header);
+    }
+
+    #[tokio::test]
+    async fn request_range_test() {
+        let (store, _) = gen_filled_store(10);
+        let expected_headers = [
+            store.get_by_height(5).unwrap(),
+            store.get_by_height(6).unwrap(),
+            store.get_by_height(7).unwrap(),
+        ];
+        let mut handler = ExchangeServerHandler::new(Arc::new(store));
+        let (mut sender, receiver) = create_test_response_sender();
+
+        let request = HeaderRequest {
+            data: Some(Data::Origin(5)),
+            amount: u64::try_from(expected_headers.len()).unwrap(),
+        };
+        handler.on_request_received(PeerId::random(), "test", request, ());
+
+        let received = select! {
+            _ = poll_fn(move |cx| handler.poll(cx, &mut sender)) => panic!("shouldn't return"),
+            r = receiver => { r.unwrap() }
+        };
+
+        for (rec, exp) in received.iter().zip(expected_headers.iter()) {
+            assert_eq!(rec.status_code, i32::from(StatusCode::Ok));
+            let decoded_header = ExtendedHeader::decode(&rec.body[..]).unwrap();
+            assert_eq!(&decoded_header, exp);
+        }
+    }
+
+    #[tokio::test]
+    async fn request_range_beyond_head_test() {
+        let (store, _) = gen_filled_store(5);
+        let expected_hashes = [store.get_by_height(5).ok()];
+        let expected_status_codes = [StatusCode::Ok];
+        assert_eq!(expected_hashes.len(), expected_status_codes.len());
+
+        let mut handler = ExchangeServerHandler::new(Arc::new(store));
+        let (mut sender, receiver) = create_test_response_sender();
+
+        let request = HeaderRequest::with_origin(5, u64::try_from(expected_hashes.len()).unwrap());
```

Comment:
I don't think this requests beyond, it looks like a request for a single header

Reply:
true, missed that request uses len from expected response. Fixed to request const `10` :+1: